scripts/download_checkpoints.py

```
@@ -0,0 +1,137 @@
+
+def try_get_directory_listing(url):
+    common_files = [
+        "config.yaml",
+        "model.pt",
+        "optim.pt",
+        "train.pt",
+        "model.safetensors",
+        "optim.safetensors",
+    ]
+    found_files = []
+    for pattern in common_files:
+        try:
+            test_url = urljoin(url.rstrip('/') + '/', pattern)
+            response = requests.head(test_url)
+            # response.raise_for_status()
+            if response.status_code == 200:
+                found_files.append(pattern)
+        except requests.exceptions.HTTPError as e:
+            print(f"HTTP error for {pattern}: {e}")
+        except requests.exceptions.RequestException as e:
+            print(f"Connection error for {pattern}: {e}")
+    return found_files
+
+def download_checkpoint(url, save_dir):
+   """Download all files from a checkpoint directory."""
+   r2_url = convert_to_r2_url(url)
+   public_url = convert_to_public_url(r2_url)
+   base_path = Path(save_dir)
+   base_path.mkdir(parents=True, exist_ok=True)
+   print(f"Saving to: {base_path}")
+   available_files = try_get_directory_listing(public_url)
+   
+   if not available_files:
+       raise ValueError("No matching files found in directory")
+   
+   failed_files = []
+   for file in available_files:
+       file_url = urljoin(public_url.rstrip('/') + '/', file)
+       file_path = base_path / file
+       try:
+           print(f"\nDownloading: {file}")
+           download_file(file_url, file_path)
+       except requests.exceptions.Timeout:
+           print(f"Timeout error for {file}, retrying once...")
+           try:
+               download_file(file_url, file_path)
+           except requests.exceptions.RequestException as e:
+               failed_files.append(file)
+               print(f"Failed to download {file}: {e}")
+       except requests.exceptions.RequestException as e:
+           failed_files.append(file)
+           print(f"Failed to download {file}: {e}")
```

Don't retry in this way. The Python requests library already supports retries. In this code we should just call `download_file()`. If a file fails, we fail the whole process (i.e., we let the exception propagate).